sbi/diagnostics/misspecification.py

```diff
@@ -0,0 +1,137 @@
+       needed to compute the distribution of mmds under the null hypothesis
+       that synthetic and observed samples come from the same distribution.
+    n_obs: number of observed data points,
+        used to determine the number of samples for one set
+    y: synthetic data
+    n_shuffle: number of shuffles
+    max_samples: maximum number of samples to use
+    mode: mode of MMD calculation
+    """
+    mmds = torch.zeros(n_shuffle)
+    if n_obs > y.shape[0]:
+        raise ValueError(
+            "n of observed samples should be less than n of synthetic samples"
+        )
+    for i in range(n_shuffle):
+        idx = torch.randperm(y.shape[0])[:max_samples]
+        mmds[i] = compute_rbf_mmd_median_heuristic(
+            y[idx[:n_obs]], y[idx[n_obs:]], mode=mode
+        )
+    return mmds
+
+
+def calculate_p_misspecification(
+    x_obs, x, n_shuffle=1_000, max_samples=1_000, mode="biased"
+):
+    """calculate the p-value of the misspecification test.
+    x_obs: observed data
+    x: synthetic data
+    n_shuffle: number of shuffles
+    max_samples: maximum number of samples to use
+    mode: mode of MMD calculation ("biased" or "unbiased")
+    """
+    mmds_baseline = calculate_baseline_mmd(
+        x_obs.shape[0], x, n_shuffle=n_shuffle, max_samples=max_samples, mode=mode
+    )
+    mmd = compute_rbf_mmd_median_heuristic(x_obs, x[:max_samples], mode=mode)
+    p_val = 1 - (mmds_baseline < mmd).sum().item() / n_shuffle
+    return p_val, (mmds_baseline, mmd)
+
+
+def calc_misspecification_mmd(
+    x_obs,
+    x,
+    inference=None,
+    mode="x_space",
+    n_shuffle=1_000,
+    max_samples=1_000,
+    mmd_mode="biased",
+):
+    """misspecification test based on MMD in data- or embedding space.
+    x_obs: observed data
+    x: synthetic data
+    inference: sbi inference object (only used if mode == "embedding")
+    mode: space of MMD calculation ("x_space" or "embedding")
+    n_shuffle: number of shuffles for computing mmds under H_0
+    max_samples: maximum number of samples to use
+        (when we have too many synthetic samples x)
+    mmd_mode: approximation of MMD calculation ("biased" or "unbiased")
+    returns:
+        p_val, (mmd_baseline,mmd): p-value of the misspecification test
+                                    (MMDs under H_0, mmd)
+    """
+    if mode == "x_space":
+        z_obs = x_obs
+        z = x
+    elif mode == "embedding":
+        if inference is None:
+            raise ValueError(
+                "inference should not be None if mode is 'embedding'."
+                "please provide an sbi inference object"
+            )
+        if not hasattr(inference, '_neural_net'):
+            raise ValueError(
+                "no neural net found,"
+                "neural_net should not be None when mode is 'embedding'"
+            )
+        if isinstance(inference._neural_net.embedding_net, nn.modules.linear.Identity):
+            raise Warning(
```

Comment:
Can you also change this to an error with a more informative message, otherwise change this to a `warn` instead of `raise Warning`